src/main/java/uk/gov/ida/stub/idp/repositories/SessionRepositoryBase.java

```diff
@@ -0,0 +1,133 @@
+package uk.gov.ida.stub.idp.repositories;
+
+import com.fasterxml.jackson.annotation.JsonAutoDetect;
+import com.fasterxml.jackson.annotation.PropertyAccessor;
+import com.fasterxml.jackson.core.JsonParseException;
+import com.fasterxml.jackson.core.JsonProcessingException;
+import com.fasterxml.jackson.databind.JsonMappingException;
+import com.fasterxml.jackson.databind.ObjectMapper;
+import com.fasterxml.jackson.datatype.joda.JodaModule;
+import org.jdbi.v3.core.Jdbi;
+import org.opensaml.core.xml.XMLObject;
+import org.opensaml.saml.saml2.core.AuthnContextComparisonTypeEnumeration;
+import uk.gov.ida.common.SessionId;
+import uk.gov.ida.saml.hub.domain.IdaAuthnRequestFromHub;
+import uk.gov.ida.stub.idp.exceptions.SessionSerializationException;
+import uk.gov.ida.stub.idp.repositories.jdbc.mixins.AuthnContextComparisonTypeMixin;
+import uk.gov.ida.stub.idp.repositories.jdbc.mixins.IdaAuthnRequestFromHubMixin;
+import uk.gov.ida.stub.idp.repositories.jdbc.mixins.XmlObjectMixin;
+
+import java.io.IOException;
+import java.time.Instant;
+import java.util.Optional;
+
+import static com.fasterxml.jackson.databind.DeserializationFeature.FAIL_ON_UNKNOWN_PROPERTIES;
+import static org.apache.commons.lang3.StringEscapeUtils.unescapeJson;
+
+public abstract class SessionRepositoryBase<T extends Session> implements SessionRepository<T> {
+    private final Jdbi jdbi;
+    private final ObjectMapper objectMapper;
+    private final Class<T> sessionType;
+
+    public SessionRepositoryBase(Class<T> sessionType, Jdbi jdbi) {
+        this.sessionType = sessionType;
+        this.jdbi = jdbi;
+        this.objectMapper = createObjectMapperForSerialization();
+    }
+    
+    public boolean containsSession(SessionId sessionToken) {
+        return jdbi.withHandle(handle -> handle.select(
+                "select count(1) from stub_idp_session where session_id = ?", sessionToken.toString())
+                .mapTo(Boolean.class)
+                .findOnly());
+    }
+    
+    public Optional<T> get(SessionId sessionToken) {
+        Optional<String> sessionData = jdbi.withHandle(handle -> handle.select(
+                "select session_data from stub_idp_session where session_id = ?", sessionToken.toString())
+                .mapTo(String.class)
+                .findFirst());
+
+        if (!sessionData.isPresent()) {
+            return Optional.empty();
+        }
+
+        try {
+            String serializedSession = sessionData.get();
+            T session = objectMapper.readValue(unescapeJson(serializedSession.substring(1, serializedSession.length() - 1)), sessionType);
```

Comment:
Rather than saving the class as a field from the constructor, it might make more sense to move the deserialization into the concrete classes as this is the major point at which their behaviour is different. Maybe adding an abstract method something like `deserializeSession(String serializedSession)` where each concrete class can do the object mapping.